addon/scripts/h2c.js

```diff
@@ -0,0 +1,204 @@
+const FFp = p256.field;
+
+/**
+ * hashes bits to the base field (as described in
+ * draft-irtf-cfrg-hash-to-curve)
+ * @param {sjcl.bitArray} x bits of element to be translated
+ * @param {sjcl.ecc.curve} curve elliptic curve
+ * @param {string} label context label for domain separation
+ * @return {int} integer in the base field of curve
+ */
+function h2Base(x, curve, label) {
+  const curveName = sjcl.ecc.curveName(curve);
+  const h = getHash(curveName);
+  h.update("h2c");
+  h.update(label);
+  h.update(x);
+  const t = h.finalize();
+  const y = curve.field.fromBits(t).cnormalize();
+  return y;
+}
+
+/**
+ * hashes bits to the chosen elliptic curve
+ * @param {sjcl.bitArray} alpha bits to be encoded onto curve
+ * @param {sjcl.ecc.curve} curve elliptic curve
+ * @param {int} mode indicates which algorithm mode should be used (if any)
+ * @return {sjcl.ecc.point} point on curve
+ */
+function h2Curve(alpha, curve, mode) {
+  const curveName = sjcl.ecc.curveName(curve);
+  let point;
+  switch (curveName) {
+    case "c256":
+      point = simplifiedSWU(alpha, mode);
+      break;
+    default:
+      throw new Error("[privacy-pass]: Incompatible curve chosen for hashing: " + curveName);
+  }
+  return point;
+}
+
+/**
+ * hashes bits onto affine P256 point using simplified SWU encoding algorithm
+ * @param {sjcl.bitArray} alpha bits to be encoded
+ * @param {int} mode 0 = affine, 1 = projective
+ * @return {sjcl.ecc.point} point on P256
+ */
+function simplifiedSWU(alpha, mode) {
+  const params = getP256Params();
+  const p = params.p;
+  const A = params.A;
+  const B = params.B;
+  const t = h2Base(alpha, p256, "p256_hashing");
+
+  let point;
+  switch (mode) {
+    case 0:
+      point = affineSWUP256(p, A, B, t);
+      break;
+    case 1:
+      point = jacobianSWUP256(p, A, B, t);
+      break;
+    default:
+      throw new Error("[privacy-pass]: Incompatible mode type chosen for SWU");
+  }
+
+  if (!point.isValid()) {
+    throw new Error("[privacy-pass]: Generated point is not on curve");
+  }
+
+  return point;
+}
+
+/**
+ * Converts field elements into an affine pair of coordinates on P256
+ * @param {sjcl.curve.field.modulus} p curve modulus
+ * @param {sjcl.bn} A Weierstrass coefficient for P256 curve
+ * @param {sjcl.bn} B (see above)
+ * @param {sjcl.bn} t FF_p field element
+ * @return {sjcl.ecc.point} point on P256
+ */
+function affineSWUP256(p, A, B, t) {
+  // step 1
+  let c = t.square();
+  c = c.mul(-1);
+
+  // step 2
+  let right = c.square().add(c);
+  right = right.inverseMod(p);
+  right = right.add(1);
+  const invA = A.inverseMod(p);
+  let X = new FFp(B.mul(invA).mul(-1).mul(right));
+
+  // step 3
+  const g = X.mul(X.square().add(A)).add(B);
+
+  // step 4
+  const expo = p.add(1).cnormalize().halveM().halveM();
+  let Y = new FFp(g.power(expo));
+
+  // step 5-8
+  const Ysq = Y.square();
+  const b = g.equals(Ysq);
+  const cX = c.mul(X);
+  const t3Y = t.power(3).mul(Y); // p+1/4 is even for P256
```

Comment:
Why is this the correct y value? It isn't clear to me from reading the RFC. If there is some algebra involved in that, it would be nice to have an explanation to make the correctness clear.